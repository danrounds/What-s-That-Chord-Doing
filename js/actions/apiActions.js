import { getReqUserScores, getReqHighScores, postReqAccount,
         putReqUserScores, putReqAccountPassword, deleteReqAccount }
from '../apiCalls';

const getScores = {
    getUserScores: (name, password) =>
        dispatch => {
            dispatch(getScores.getUserScoresPending());
            return getReqUserScores(name, password)
                .then(scores => dispatch(getScores.getUserScoresSuccess(scores)))
                .catch((e) => dispatch(getScores.getUserScoresFailure(e)));
        },

    GET_USER_SCORES_PENDING: 'GET_USER_SCORES_PENDING',
    getUserScoresPending: () => ({
        type: GET_USER_SCORES_PENDING
    }),

    GET_USER_SCORES_SUCCESS: 'GET_USER_SCORES_SUCCESS',
    getUserScoresSuccess: (scores) => ({
        type: GET_USER_SCORES_SUCCESS,
        scores
    }),

    GET_USER_SCORES_FAILURE: 'GET_USER_SCORES_FAILURE',
    getUserScoresFailure: (error) => ({
        type: GET_USER_SCORES_FAILURE,
        error
    }),
};

export const LOG_OFF = 'LOG_OFF';
export const logOff = () => ({
    type: LOG_OFF
});

const updateScores = {
    updateUserScores: (name, password, scores) =>
        dispatch => putReqUserScores(name, password, scores)
        .then(() => dispatch(updateScores.updateUserScoresSuccess(scores)))
        .catch((e) => dispatch(updateScores.updateUserScoresFailure(e))),

    UPDATE_USER_SCORES_SUCCESS: 'UPDATE_USER_SCORES_SUCCESS',
    updateUserScoresSuccess: () => ({
        type: UPDATE_USER_SCORES_SUCCESS
    }),

    UPDATE_USER_SCORES_FAILURE: 'UPDATE_USER_SCORES_FAILURE',
    updateUserScoresFailure: (error) => ({
        type: UPDATE_USER_SCORES_FAILURE,
        error
    }),
};

const getHighScores_ = {
    getHighScores: (gameType) =>
        dispatch => {
            dispatch(getHighScores_.getHighScoresPending());
            return getReqHighScores(gameType)
                .then(highScores => dispatch(getHighScores_.getHighScoresSuccess(highScores)))
                .catch((e) => dispatch(getHighScores_.getHighScoresFailure(e)));
        },

    GET_HIGH_SCORES_PENDING: 'GET_HIGH_SCORES_PENDING',
    getHighScoresPending: () => ({
        type: GET_HIGH_SCORES_PENDING
    }),

    GET_HIGH_SCORES_SUCCESS: 'GET_HIGH_SCORES_SUCCESS',
    getHighScoresSuccess: (highScores) => ({
        type: GET_HIGH_SCORES_SUCCESS,
        highScores
    }),

    GET_HIGH_SCORES_FAILURE: 'GET_HIGH_SCORES_FAILURE',
    getHighScoresFailure: (error) => ({
        type: GET_HIGH_SCORES_FAILURE,
        error
    }),
}

const makeAccount = {
    makeUserAccount: (name, password) =>
        dispatch => postReqAccount(name, password)
        .then(() => dispatch(makeAccount.makeUserAccountSuccess()))
        .catch(() => dispatch(makeAccount.makeUserAccountFailure())),

    MAKE_USER_ACCOUNT_SUCCESS: 'MAKE_USER_ACCOUNT_SUCCESS',
    makeUserAccountSuccess: () => ({
        type: MAKE_USER_ACCOUNT_SUCCESS
    }),

    MAKE_USER_ACCOUNT_FAILURE: 'MAKE_USER_ACCOUNT_FAILURE',
    makeUserAccountFailure: (error) => ({
        type: MAKE_USER_ACCOUNT_FAILURE,
        error
    }),
};

const changePass = {
    changeUserPassword: (name, password) =>
        dispatch => putReqAccountPassword(name, password)
        .then(() => dispatch(changePass.changeUserPasswordSuccess()))
        .catch((e) => dispatch(changePass.changeUserPasswordFailure())),

    CHANGE_USER_PASSWORD_SUCCESS: 'CHANGE_USER_PASSWORD_SUCCESS',
    changeUserPasswordSuccess: () => ({
        type: CHANGE_USER_PASSWORD_SUCCESS
    }),

    CHANGE_USER_PASSWORD_FAILURE: 'CHANGE_USER_PASSWORD_FAILURE',
    changeUserPasswordFailure: (error) => ({
        type: CHANGE_USER_PASSWORD_FAILURE,
        error
    }),
};

///
///
// I've not bothered to write an action for account deletion, because that
// doesn't belong in the MVP client

// We end up with a bit more code by grouping associated actions together, but
// it at least tells us which goes with which
export const { getUserScores, GET_USER_SCORES_PENDING, getUserScoresPending,
               GET_USER_SCORES_SUCCESS, getUserScoresSuccess,
               GET_USER_SCORES_FAILURE, getUserScoresFailure } = getScores;

export const { updateUserScores, UPDATE_USER_SCORES_SUCCESS,
               updateUserScoresSuccess, UPDATE_USER_SCORES_FAILURE,
               updateUserScoresFailure } = updateScores;

export const { getHighScores, GET_HIGH_SCORES_PENDING, getHighScoresPending,
               GET_HIGH_SCORES_SUCCESS, getHighScoresSuccess,
               GET_HIGH_SCORES_FAILURE, getHighScoresFailure } = getHighScores_;

export const { makeUserAccount, MAKE_USER_ACCOUNT_SUCCESS,
               makeUserAccountSuccess, MAKE_USER_ACCOUNT_FAILURE,
               makeUserAccountFailure } = makeAccount;

export const { changeUserPassword, CHANGE_USER_PASSWORD_SUCCESS,
               changeUserPasswordSuccess, CHANGE_USER_PASSWORD_FAILURE,
               changeUserPasswordFailure } = changePass;
