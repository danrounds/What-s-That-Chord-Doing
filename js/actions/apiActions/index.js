export * from './actionsAccountAndPassword';
export * from './actionsLogInOff';
export * from './actionsScores';
