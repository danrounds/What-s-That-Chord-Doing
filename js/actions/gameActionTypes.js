export const SET_KEY_PRESS = 'SET_KEY_PRESS';
export const START_NEW_GAME = 'START_NEW_GAME';
export const GIVE_UP = 'GIVE_UP';
export const GET_NEXT_QUESTION = 'GET_NEXT_QUESTION';
export const MAKE_GUESS = 'MAKE_GUESS';
export const MARK_TURN_CORRECT = 'MARK_TURN_CORRECT';
