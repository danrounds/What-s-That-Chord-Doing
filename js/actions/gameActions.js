import chordGetter from '../musicLogic/';

export const SET_KEY_PRESS = 'SET_KEY_PRESS';
export const setKeyPress = (keyValue) => ({
    type: SET_KEY_PRESS,
    keyValue,
});

export const START_NEW_GAME = 'START_NEW_GAME';
export const startNewGame = (gameType, inversions) => {
    let {keyNameReadable, keyNameNotation, introChordSequence, chordSubset} =
            chordGetter.init(gameType, inversions);
    let {chordName, currentChordNumeral, bassNote, trebleNotes, accidentals} =
            chordGetter.getChord();
    let gameNumber = Math.random() * Math.pow(2,52);
    // ^ This is not a seed number; it's merely an [ostensibly] unique
    // identifier, that we'll use to decide whether components should update

    return {
        type: START_NEW_GAME,
        gameType,
        gameNumber,
        inversions,
        keyNameReadable,
        keyNameNotation,
        introChordSequence,
        chordSubset,
        chordName,
        currentChordNumeral,
        bassNote,
        trebleNotes,
        accidentals
    };
};
export const GIVE_UP = 'GIVE_UP';
export const giveUp = () => ({
    type: GIVE_UP,
});

export const GET_NEXT_QUESTION = 'GET_NEXT_QUESTION';
export const getNextQuestion = () => {
    let {chordName, currentChordNumeral, bassNote, trebleNotes, accidentals} =
            chordGetter.getChord();
    return {
        type: GET_NEXT_QUESTION,
        chordName,
        currentChordNumeral,
        bassNote,
        trebleNotes,
        accidentals,
    };
};

export const MAKE_GUESS = 'MAKE_GUESS';
export const makeGuess = (guess) => ({
    type: MAKE_GUESS,
    guess,
});

export const MARK_TURN_CORRECT = 'MARK_TURN_CORRECT';
export const markTurnCorrect = () => ({
    type: MARK_TURN_CORRECT,
});
