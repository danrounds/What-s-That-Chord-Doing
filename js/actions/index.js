import chordGetter from '../music-logic/';

export const GET_KEY_PRESS = 'GET_KEY_PRESS';
export const getKeyPress = (keyValue) => ({
    type: GET_KEY_PRESS,
    keyValue
});

export const TOGGLE_KEYBOARD_SHORTCUT_DISPLAY = 'TOGGLE_KEYBOARD_SHORTCUT_DISPLAY';
export const toggleKeyboardShortcutDisplay = (displayBool) => ({
    type: TOGGLE_KEYBOARD_SHORTCUT_DISPLAY,
    displayBool
});


export const START_NEW_GAME = 'START_NEW_GAME';
export const startNewGame = (gameType, inversions) => {
    let {keyNameReadable, keyNameNotation, introChordSequence, chordSubset} = chordGetter.init(gameType, inversions);
    let {chordName, currentChordNumeral, bassNote, trebleNotes, accidentals} = chordGetter.getChord();
    return {
        type: START_NEW_GAME,
        inversions,
        gameType,
        keyNameReadable,
        keyNameNotation,
        introChordSequence,
        chordSubset,
        chordName,
        currentChordNumeral,
        bassNote,
        trebleNotes,
        accidentals
    };
};

export const GET_NEXT_QUESTION = 'GET_NEXT_QUESTION';
export const getNextQuestion = () => {
    let {chordName, currentChordNumeral, bassNote, trebleNotes, accidentals} = chordGetter.getChord();
    return {
        type: GET_NEXT_QUESTION,
        chordName,
        currentChordNumeral,
        bassNote,
        trebleNotes,
        accidentals
    };
};

export const INCREMENT_GUESS_N = 'INCREMENT_GUESS_N';
export const incrementGuessN = () => ({
    type: INCREMENT_GUESS_N
});

export const MARK_TURN_CORRECT = 'MARK_TURN_CORRECT';
export const markTurnCorrect = () => ({
    type: MARK_TURN_CORRECT
});
