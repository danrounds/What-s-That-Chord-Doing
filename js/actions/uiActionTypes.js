export const UPDATE_LESSON_INDEX_DISPLAY = 'UPDATE_LESSON_INDEX_DISPLAY';
export const TOGGLE_KEYBOARD_SHORTCUT_DISPLAY = 'TOGGLE_KEYBOARD_SHORTCUT_DISPLAY';
