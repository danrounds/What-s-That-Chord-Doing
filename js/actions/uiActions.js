export const UPDATE_LESSON_INDEX_DISPLAY = 'UPDATE_LESSON_INDEX_DISPLAY';
export const updateLessonIndexDisplay = (keyValueToUpdate) => ({
    type: UPDATE_LESSON_INDEX_DISPLAY,
    keyValueToUpdate,
});

export const TOGGLE_KEYBOARD_SHORTCUT_DISPLAY = 'TOGGLE_KEYBOARD_SHORTCUT_DISPLAY';
export const toggleKeyboardShortcutDisplay = (displayBool) => {
    localStorage.setItem('@WTCD/displayKeyboardShortcuts', displayBool);
    return {
        type: TOGGLE_KEYBOARD_SHORTCUT_DISPLAY,
        displayBool,
    };
};
