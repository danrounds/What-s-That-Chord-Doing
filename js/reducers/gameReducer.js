import * as actions from '../actions/gameActions';

const initialState = {
    keyValue: null,
    gameType: null,
    gameNumber: null,
    inversions: null,
    keyNameReadable: null,
    keyNameNotation: null,
    introChordSequence: [ [],[],[],[] ],
    chordSubset: [],
    chordName: null,
    chord: null,
    notes: { bass: null, treble: [] },
    inversionN: null,
    accidentals: { bassAccidental: null, trebleIndices: [] },
    guess: null,
    guessN: 0,
    answeredCorrectly: null,
    giveUp: null,
    questionNumber: 0,
    nAnsweredRight: 0,
    clicksPerRightAnswer: [],
    gameOver: false,
};

// Our reducer for the game-play/keyboard-shortcut part of our app:
export default (state=initialState, action) => {

    switch(action.type) {

    case actions.SET_KEY_PRESS:
        return {
            keyValue: action.keyValue,
            gameType: state.gameType,
            gameNumber: state.gameNumber,
            inversions: state.inversions,
            keyNameReadable: state.keyNameReadable,
            keyNameNotation: state.keyNameNotation,
            introChordSequence: [[...state.introChordSequence[0]],
                                 [...state.introChordSequence[1]],
                                 [...state.introChordSequence[2]],
                                 [...state.introChordSequence[3]]],
            chordSubset: [ ...state.chordSubset ],
            chordName: state.chordName,
            chord: state.chord,
            notes: {
                bass: state.notes.bass,
                treble: [...state.notes.treble]
            },
            inversionN: state.inversionN,
            accidentals: {
                bassAccidental: state.accidentals.bassAccidental,
                trebleIndices: [...state.accidentals.trebleIndices],
            },
            guess: state.guess,
            guessN: state.guessN,
            answeredCorrectly: state.answeredCorrectly,
            giveUp: state.giveUp,
            questionNumber: state.questionNumber,
            nAnsweredRight: state.nAnsweredRight,
            clicksPerRightAnswer: [...state.clicksPerRightAnswer],
            gameOver: state.gameOver
        };

    case actions.START_NEW_GAME:
        return {
            keyValue: '',
            gameType: action.gameType,
            gameNumber: action.gameNumber,
            inversions: action.inversions,
            keyNameReadable: action.keyNameReadable,
            keyNameNotation: action.keyNameNotation,
            introChordSequence: action.introChordSequence,
            chordSubset: action.chordSubset,
            chordName: action.chordName,
            chord: action.currentChordNumeral,
            notes: { bass: action.bassNote, treble: action.trebleNotes },
            inversionN: action.inversionN,
            accidentals: action.accidentals,
            guess: null,
            guessN: 0,
            answeredCorrectly: false,
            giveUp: false,
            questionNumber: 1,
            nAnsweredRight: 0,
            clicksPerRightAnswer: [],
            gameOver: false
        };

    case actions.GIVE_UP:
        return {
            keyValue: '',
            gameType: state.gameType,
            gameNumber: state.gameNumber,
            inversions: state.inversions,
            keyNameReadable: state.keyNameReadable,
            keyNameNotation: state.keyNameNotation,
            introChordSequence: [[...state.introChordSequence[0]],
                                 [...state.introChordSequence[1]],
                                 [...state.introChordSequence[2]],
                                 [...state.introChordSequence[3]]],
            chordSubset: [...state.chordSubset],
            chordName: state.chordName,
            chord: state.chord,
            notes: {
                bass: state.notes.bass,
                treble: [...state.notes.treble]
            },
            inversionN: state.inversionN,
            accidentals: {
                bassAccidental: state.accidentals.bassAccidental,
                trebleIndices: [...state.accidentals.trebleIndices]
            },
            guess: null,
            guessN: 0,
            answeredCorrectly: false,
            giveUp: true,
            questionNumber: state.questionNumber,
            nAnsweredRight: state.nAnsweredRight,
            clicksPerRightAnswer: [...state.clicksPerRightAnswer],
            gameOver: state.questionNumber === 10 ? true : false,
        };

    case actions.GET_NEXT_QUESTION:
        return {
            keyValue: '',
            gameType: state.gameType,
            gameNumber: state.gameNumber,
            inversions: state.inversions,
            keyNameReadable: state.keyNameReadable,
            keyNameNotation: state.keyNameNotation,
            introChordSequence: [[...state.introChordSequence[0]],
                                 [...state.introChordSequence[1]],
                                 [...state.introChordSequence[2]],
                                 [...state.introChordSequence[3]]],
            chordSubset: [...state.chordSubset],
            chordName: action.chordName,
            chord: action.currentChordNumeral,
            notes: { bass: action.bassNote, treble: [...action.trebleNotes] },
            inversionN: action.inversionN,
            accidentals: {
                bassAccidental: action.accidentals.bassAccidental,
                trebleIndices: [...action.accidentals.trebleIndices]
            },
            guess: null,
            guessN: 0,
            answeredCorrectly: false,
            giveUp: false,
            questionNumber: (state.questionNumber + 1) % 11,
            nAnsweredRight: state.nAnsweredRight,
            clicksPerRightAnswer: [...state.clicksPerRightAnswer],
            gameOver: false
        };

    case actions.MAKE_GUESS:
        return {
            keyValue: '',
            gameType: state.gameType,
            gameNumber: state.gameNumber,
            inversions: state.inversions,
            keyNameReadable: state.keyNameReadable,
            keyNameNotation: state.keyNameNotation,
            introChordSequence: [[...state.introChordSequence[0]],
                                 [...state.introChordSequence[1]],
                                 [...state.introChordSequence[2]],
                                 [...state.introChordSequence[3]]],
            chordSubset: [...state.chordSubset],
            chordName: state.chordName,
            chord: state.chord,
            notes: {
                bass: state.notes.bass,
                treble: [...state.notes.treble]
            },
            inversionN: state.inversionN,
            accidentals: {
                bassAccidental: state.accidentals.bassAccidental,
                trebleIndices: [...state.accidentals.trebleIndices]
            },
            guess: action.guess,
            guessN: state.guessN + 1,
            answeredCorrectly: false,
            giveUp: state.giveUp,
            questionNumber: state.questionNumber,
            nAnsweredRight: state.nAnsweredRight,
            clicksPerRightAnswer: [...state.clicksPerRightAnswer],
            gameOver: false
        };

    case actions.MARK_TURN_CORRECT:
        return {
            keyValue: '',
            gameType: state.gameType,
            gameNumber: state.gameNumber,
            inversions: state.inversions,
            keyNameReadable: state.keyNameReadable,
            keyNameNotation: state.keyNameNotation,
            introChordSequence: [[...state.introChordSequence[0]],
                                 [...state.introChordSequence[1]],
                                 [...state.introChordSequence[2]],
                                 [...state.introChordSequence[3]]],
            chordSubset: [...state.chordSubset],
            chordName: state.chordName,
            chord: state.chord,
            notes: {
                bass: state.notes.bass,
                treble: [...state.notes.treble]
            },
            inversionN: state.inversionN,
            accidentals: {
                bassAccidental: state.accidentals.bassAccidental,
                trebleIndices: [...state.accidentals.trebleIndices]
            },
            guess: state.guess,
            guessN: state.guessN,
            answeredCorrectly: true,
            questionNumber: state.questionNumber,
            nAnsweredRight: state.nAnsweredRight + 1,
            clicksPerRightAnswer: [...state.clicksPerRightAnswer, state.guessN],
            gameOver: state.questionNumber === 10 ? true : false,
        };

    default:
        return state;
    }
};
