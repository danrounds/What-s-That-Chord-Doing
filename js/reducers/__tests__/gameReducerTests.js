import gameReducer from '../gameReducer';
import * as actions from '../../actions/gameActions';

actions.startNewGame = (gameType, inversions) => ({
    type: 'START_NEW_GAME',
    gameType,
    gameNumber: 392549214,
    inversions,
    keyNameReadable: 'Ab minor',
    keyNameNotation: 'Abm',
    introChordSequence: [
        ['G#/3','D#/5','G#/5','B/5',],
        ['C#/4','C#/5','E/5','G#/5',],
        ['D#/4','A#/4','D#/5','G/5',],
        ['G#/3','B/4','D#/5','G#/5',]
    ],
    chordSubset: [
        'i','♭II','ii°','ii','♭III','iv','v','V','♭VI','vi','♭VII','vii°',
    ],
    chordName: 'G diminished',
    currentChordNumeral: 'vii°',
    bassNote: 'G/4',
    trebleNotes: ['G/5','Bb/5','Db/6',],
    inversionN: 0,
    accidentals: {
        bassAccidental: true,
        trebleIndices: [0]
    },
});

actions.getNextQuestion = () => ({
    type: 'GET_NEXT_QUESTION',
    chordName: 'Cb major',
    currentChordNumeral:'♭III',
    bassNote: 'Cb/4',
    trebleNotes: ['Cb/5','Eb/5','Gb/5',],
    inversionN: 1,
    accidentals: {
        bassAccidental: false,
        trebleIndices: [],
    },
});

describe('gameReducer -- the reducer for our game actions', () => {
    let state;
    beforeEach(() => state = gameReducer(undefined, {}));

    it('should have a consistent initial state', () => {
        expect(state).toMatchSnapshot();
   });

    it('should register key presses', () => {
        state = gameReducer(state, actions.setKeyPress('Enter'));
        expect(state).toMatchSnapshot();
    });

    it('should start a new game', () => {
        state = gameReducer(state, actions.startNewGame('easyMinor', false));
        expect(state).toMatchSnapshot();
    });

    it('should let us give up, when we don\'t know the answer', () => {
        state = gameReducer(state, actions.startNewGame('easyMinor', false));
        state = gameReducer(state, actions.giveUp());
        expect(state).toMatchSnapshot();
    });

    it('get the next question', () => {
        state = gameReducer(state, actions.startNewGame('easyMinor', false));
        state = gameReducer(state, actions.getNextQuestion());
        expect(state).toMatchSnapshot();
    });

    it('make a guess', () => {
        state = gameReducer(state, actions.startNewGame('easyMinor', false));
        state = gameReducer(state, actions.makeGuess('iii'));
        expect(state).toMatchSnapshot();
    });
    
    it('mark turn correct when the game isn\'t about to be over', () => {
        state = gameReducer(state, actions.startNewGame('easyMinor', false));
        state = gameReducer(state, actions.markTurnCorrect());
        expect(state).toMatchSnapshot();
    });
    it('mark turn correct when the game is about to be over', () => {
        state = gameReducer(state, actions.startNewGame('easyMinor', false));
        state = gameReducer(state, actions.makeGuess('iii'));
        state = gameReducer(Object.assign(state, { questionNumber: 10 }),
                            actions.markTurnCorrect());
        expect(state).toMatchSnapshot();
    });
    

});
