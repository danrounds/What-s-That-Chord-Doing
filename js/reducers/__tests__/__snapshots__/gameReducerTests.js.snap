// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameReducer -- the reducer for our game actions get the next question 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": false,
    "trebleIndices": Array [],
  },
  "answeredCorrectly": false,
  "chord": "♭III",
  "chordName": "Cb major",
  "chordSubset": Array [
    "i",
    "♭II",
    "ii°",
    "ii",
    "♭III",
    "iv",
    "v",
    "V",
    "♭VI",
    "vi",
    "♭VII",
    "vii°",
  ],
  "clicksPerRightAnswer": Array [],
  "gameNumber": 392549214,
  "gameOver": false,
  "gameType": "easyMinor",
  "giveUp": false,
  "guess": null,
  "guessN": 0,
  "introChordSequence": Array [
    Array [
      "G#/3",
      "D#/5",
      "G#/5",
      "B/5",
    ],
    Array [
      "C#/4",
      "C#/5",
      "E/5",
      "G#/5",
    ],
    Array [
      "D#/4",
      "A#/4",
      "D#/5",
      "G/5",
    ],
    Array [
      "G#/3",
      "B/4",
      "D#/5",
      "G#/5",
    ],
  ],
  "inversionN": 1,
  "inversions": false,
  "keyNameNotation": "Abm",
  "keyNameReadable": "Ab minor",
  "keyValue": "",
  "nAnsweredRight": 0,
  "notes": Object {
    "bass": "Cb/4",
    "treble": Array [
      "Cb/5",
      "Eb/5",
      "Gb/5",
    ],
  },
  "questionNumber": 2,
}
`;

exports[`gameReducer -- the reducer for our game actions make a guess 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": true,
    "trebleIndices": Array [
      0,
    ],
  },
  "answeredCorrectly": false,
  "chord": "vii°",
  "chordName": "G diminished",
  "chordSubset": Array [
    "i",
    "♭II",
    "ii°",
    "ii",
    "♭III",
    "iv",
    "v",
    "V",
    "♭VI",
    "vi",
    "♭VII",
    "vii°",
  ],
  "clicksPerRightAnswer": Array [],
  "gameNumber": 392549214,
  "gameOver": false,
  "gameType": "easyMinor",
  "giveUp": false,
  "guess": "iii",
  "guessN": 1,
  "introChordSequence": Array [
    Array [
      "G#/3",
      "D#/5",
      "G#/5",
      "B/5",
    ],
    Array [
      "C#/4",
      "C#/5",
      "E/5",
      "G#/5",
    ],
    Array [
      "D#/4",
      "A#/4",
      "D#/5",
      "G/5",
    ],
    Array [
      "G#/3",
      "B/4",
      "D#/5",
      "G#/5",
    ],
  ],
  "inversionN": 0,
  "inversions": false,
  "keyNameNotation": "Abm",
  "keyNameReadable": "Ab minor",
  "keyValue": "",
  "nAnsweredRight": 0,
  "notes": Object {
    "bass": "G/4",
    "treble": Array [
      "G/5",
      "Bb/5",
      "Db/6",
    ],
  },
  "questionNumber": 1,
}
`;

exports[`gameReducer -- the reducer for our game actions mark turn correct when the game is about to be over 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": true,
    "trebleIndices": Array [
      0,
    ],
  },
  "answeredCorrectly": true,
  "chord": "vii°",
  "chordName": "G diminished",
  "chordSubset": Array [
    "i",
    "♭II",
    "ii°",
    "ii",
    "♭III",
    "iv",
    "v",
    "V",
    "♭VI",
    "vi",
    "♭VII",
    "vii°",
  ],
  "clicksPerRightAnswer": Array [
    1,
  ],
  "gameNumber": 392549214,
  "gameOver": true,
  "gameType": "easyMinor",
  "guess": "iii",
  "guessN": 1,
  "introChordSequence": Array [
    Array [
      "G#/3",
      "D#/5",
      "G#/5",
      "B/5",
    ],
    Array [
      "C#/4",
      "C#/5",
      "E/5",
      "G#/5",
    ],
    Array [
      "D#/4",
      "A#/4",
      "D#/5",
      "G/5",
    ],
    Array [
      "G#/3",
      "B/4",
      "D#/5",
      "G#/5",
    ],
  ],
  "inversionN": 0,
  "inversions": false,
  "keyNameNotation": "Abm",
  "keyNameReadable": "Ab minor",
  "keyValue": "",
  "nAnsweredRight": 1,
  "notes": Object {
    "bass": "G/4",
    "treble": Array [
      "G/5",
      "Bb/5",
      "Db/6",
    ],
  },
  "questionNumber": 10,
}
`;

exports[`gameReducer -- the reducer for our game actions mark turn correct when the game isn't about to be over 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": true,
    "trebleIndices": Array [
      0,
    ],
  },
  "answeredCorrectly": true,
  "chord": "vii°",
  "chordName": "G diminished",
  "chordSubset": Array [
    "i",
    "♭II",
    "ii°",
    "ii",
    "♭III",
    "iv",
    "v",
    "V",
    "♭VI",
    "vi",
    "♭VII",
    "vii°",
  ],
  "clicksPerRightAnswer": Array [
    0,
  ],
  "gameNumber": 392549214,
  "gameOver": false,
  "gameType": "easyMinor",
  "guess": null,
  "guessN": 0,
  "introChordSequence": Array [
    Array [
      "G#/3",
      "D#/5",
      "G#/5",
      "B/5",
    ],
    Array [
      "C#/4",
      "C#/5",
      "E/5",
      "G#/5",
    ],
    Array [
      "D#/4",
      "A#/4",
      "D#/5",
      "G/5",
    ],
    Array [
      "G#/3",
      "B/4",
      "D#/5",
      "G#/5",
    ],
  ],
  "inversionN": 0,
  "inversions": false,
  "keyNameNotation": "Abm",
  "keyNameReadable": "Ab minor",
  "keyValue": "",
  "nAnsweredRight": 1,
  "notes": Object {
    "bass": "G/4",
    "treble": Array [
      "G/5",
      "Bb/5",
      "Db/6",
    ],
  },
  "questionNumber": 1,
}
`;

exports[`gameReducer -- the reducer for our game actions should have a consistent initial state 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": null,
    "trebleIndices": Array [],
  },
  "answeredCorrectly": null,
  "chord": null,
  "chordName": null,
  "chordSubset": Array [],
  "clicksPerRightAnswer": Array [],
  "gameNumber": null,
  "gameOver": false,
  "gameType": null,
  "giveUp": null,
  "guess": null,
  "guessN": 0,
  "introChordSequence": Array [
    Array [],
    Array [],
    Array [],
    Array [],
  ],
  "inversionN": null,
  "inversions": null,
  "keyNameNotation": null,
  "keyNameReadable": null,
  "keyValue": null,
  "nAnsweredRight": 0,
  "notes": Object {
    "bass": null,
    "treble": Array [],
  },
  "questionNumber": 0,
}
`;

exports[`gameReducer -- the reducer for our game actions should let us give up, when we don't know the answer 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": true,
    "trebleIndices": Array [
      0,
    ],
  },
  "answeredCorrectly": false,
  "chord": "vii°",
  "chordName": "G diminished",
  "chordSubset": Array [
    "i",
    "♭II",
    "ii°",
    "ii",
    "♭III",
    "iv",
    "v",
    "V",
    "♭VI",
    "vi",
    "♭VII",
    "vii°",
  ],
  "clicksPerRightAnswer": Array [],
  "gameNumber": 392549214,
  "gameOver": false,
  "gameType": "easyMinor",
  "giveUp": true,
  "guess": null,
  "guessN": 0,
  "introChordSequence": Array [
    Array [
      "G#/3",
      "D#/5",
      "G#/5",
      "B/5",
    ],
    Array [
      "C#/4",
      "C#/5",
      "E/5",
      "G#/5",
    ],
    Array [
      "D#/4",
      "A#/4",
      "D#/5",
      "G/5",
    ],
    Array [
      "G#/3",
      "B/4",
      "D#/5",
      "G#/5",
    ],
  ],
  "inversionN": 0,
  "inversions": false,
  "keyNameNotation": "Abm",
  "keyNameReadable": "Ab minor",
  "keyValue": "",
  "nAnsweredRight": 0,
  "notes": Object {
    "bass": "G/4",
    "treble": Array [
      "G/5",
      "Bb/5",
      "Db/6",
    ],
  },
  "questionNumber": 1,
}
`;

exports[`gameReducer -- the reducer for our game actions should register key presses 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": null,
    "trebleIndices": Array [],
  },
  "answeredCorrectly": null,
  "chord": null,
  "chordName": null,
  "chordSubset": Array [],
  "clicksPerRightAnswer": Array [],
  "gameNumber": null,
  "gameOver": false,
  "gameType": null,
  "giveUp": null,
  "guess": null,
  "guessN": 0,
  "introChordSequence": Array [
    Array [],
    Array [],
    Array [],
    Array [],
  ],
  "inversionN": null,
  "inversions": null,
  "keyNameNotation": null,
  "keyNameReadable": null,
  "keyValue": "Enter",
  "nAnsweredRight": 0,
  "notes": Object {
    "bass": null,
    "treble": Array [],
  },
  "questionNumber": 0,
}
`;

exports[`gameReducer -- the reducer for our game actions should start a new game 1`] = `
Object {
  "accidentals": Object {
    "bassAccidental": true,
    "trebleIndices": Array [
      0,
    ],
  },
  "answeredCorrectly": false,
  "chord": "vii°",
  "chordName": "G diminished",
  "chordSubset": Array [
    "i",
    "♭II",
    "ii°",
    "ii",
    "♭III",
    "iv",
    "v",
    "V",
    "♭VI",
    "vi",
    "♭VII",
    "vii°",
  ],
  "clicksPerRightAnswer": Array [],
  "gameNumber": 392549214,
  "gameOver": false,
  "gameType": "easyMinor",
  "giveUp": false,
  "guess": null,
  "guessN": 0,
  "introChordSequence": Array [
    Array [
      "G#/3",
      "D#/5",
      "G#/5",
      "B/5",
    ],
    Array [
      "C#/4",
      "C#/5",
      "E/5",
      "G#/5",
    ],
    Array [
      "D#/4",
      "A#/4",
      "D#/5",
      "G/5",
    ],
    Array [
      "G#/3",
      "B/4",
      "D#/5",
      "G#/5",
    ],
  ],
  "inversionN": 0,
  "inversions": false,
  "keyNameNotation": "Abm",
  "keyNameReadable": "Ab minor",
  "keyValue": "",
  "nAnsweredRight": 0,
  "notes": Object {
    "bass": "G/4",
    "treble": Array [
      "G/5",
      "Bb/5",
      "Db/6",
    ],
  },
  "questionNumber": 1,
}
`;
